### PR TITLE
feat(api): read releases

### LINKED ISSUE #531: Service - Releases persistence & endpoints
**Context.** Releases are the immutable snapshots that pin one revision of each configurable resource. A release is created once and can then be deployed to any number of environments unchanged. This ticket introduces the release entity, its storage, and the REST endpoints for reading and assembling releases from existing revision ids.

**Description.**

Data model:

```json
{
  "id": "rel_01KX...",
  "project_id": "proj_...",
  "audit": {
    "message": "add phone_number to human-user",
    "git_sha": "4a5b6c7d8e9f...",
    "git_dirty": false,
    "created_at": "2026-07-14T09:11:00Z",
    "created_by": "user_..."
  },
  "pointers": [
    { "kind": "schema", "handle": "human-user",   "revision_id": "sch_..." },
    { "kind": "flow",   "handle": "default-login", "revision_id": "flowdef_..." }
  ]
}
```

| Field | Notes |
|---|---|
| `id` | Opaque, immutable, assigned at construction. |
| `audit.message` | Caller-supplied short summary (analogous to a git commit message). |
| `audit.git_sha` | The source commit the CLI was operating from. |
| `audit.git_dirty` | `true` when the working tree had uncommitted changes at construction — the release content does not correspond exactly to `git_sha`. |
| `audit.created_at`, `audit.created_by` | Timestamp and identity of the caller. |
| `pointers` | List of `(kind, handle, revision_id)` tuples. No resource content embedded. |

Storage shape (single table vs. join table for pointers, JSONB vs. relational) is up to the implementer.

Endpoints:

- `POST /releases` — bundles a release from existing (`kind`, `handle`, `revision_id`) tuples.
  - Validates that every referenced `revision_id` exists and that handle references inside the pointer set are internally consistent.
  - Does not allocate new revisions. 
  - Idempotent on the pointer set: same set (audit metadata excluded from the content hash) returns the existing `release_id` (200 / 409 - TBD)
- `GET /releases` — list releases in the project, newest first. Each entry carries audit metadata; pointer tuples are omitted.
- `GET /releases/{id}` — read one release: audit metadata + pointer tuples. Does not embed resource content.

**Blocks.** The bundle constructor endpoint (`POST /configuration-releases`) depends on this ticket (#545) .

**Related ADR sections.** [Releases](https://github.com/zitadel/nextgen/blob/main/docs/adrs/035-configuration-environments.md#releases), [API — `releases`](https://github.com/zitadel/nextgen/blob/main/docs/adrs/035-configuration-environments.md#releases-1).

### PR DESCRIPTION
**Pull request by vitorbari**
Implements `GET /releases` and `GET /releases/{release_id}`.

## `GET /releases`

Screenshot 2026-09-07 at 11 44 50

## `GET /releases/{release_id}`

Screenshot 2026-09-07 at 11 44 55

### CODE REVIEW

File: internal/service/release.go [hunk 1]
```diff
@@ -35,8 +36,21 @@ type CreateReleaseOutput struct {
...
+
+type ListReleasesOutput struct {
+	Items         []*domain.Release
+	NextPageToken string
+}
+
 type ReleaseService interface {
 	Create(ctx context.Context, input CreateReleaseInput) (*CreateReleaseOutput, error)
+	Get(ctx context.Context, projectID, id string) (*domain.Release, error)
+	List(ctx context.Context, input ListReleasesInput) (*ListReleasesOutput, error)
```
Comment (IAM-marco) [inline internal/service/release.go]: Past review note (confidence 100): this PR grows the `ReleaseService` interface from one method to three, and there is still an open review comment from PR #1129 asking for this type to get a proper shape first.

What changed: `Get` and `List` are added here, and both new endpoints call them through the handler:
https://github.com/zitadel/nextgen/blob/1b65bd691d683ea9178eb1a7b97cf972f7ebfead/internal/service/release.go#L50-L54

Why this spot matters: this is the only service in the package that is an interface with a single implementation and no generated mock. Every other service is either a plain struct (like `TeamService` or `BrandingService`) or an interface listed in the package's mock generation run, which this one is not:
https://github.com/zitadel/nextgen/blob/1b65bd691d683ea9178eb1a7b97cf972f7ebfead/internal/service/mocks/generate.go#L14-L16

What follows concretely: unit tests build the handler with nil in this slot today, for example here:
https://github.com/zitadel/nextgen/blob/1b65bd691d683ea9178eb1a7b97cf972f7ebfead/internal/api/flow_test.go#L120-L122
The first unit test that reaches a release path through the handler will crash with a nil panic instead of failing with a readable mock message. This PR doubles the surface where that can happen.

Separately: the PR #1129 comment explicitly predicted that these two read endpoints would land on top of the unresolved type, which is exactly what is happening here.

Fix direction: either make the service a plain struct like its siblings, or add `ReleaseService` to the mock generation list and hand the tests a mock instead of nil. With three methods now using the interface, the mock route likely pays for itself, as the #1129 comment suggested.

Coverage note: nothing fails today only because no unit test yet touches a release method through the handler, so the first one written is the one that hits the panic.

File: internal/api/release_internal_test.go [hunk 2]
```diff
@@ -60,3 +61,61 @@ func TestReleaseAccessRow(t *testing.T) {
...
+	}
+	filter, ok := service.AuthzListFilterFromContext(ctx)
+	if !ok {
+		t.Fatal("Forbidden must attach the EXISTS list filter")
+	}
+	// The kind is what scopes the filter to releases; the wrong one here would
+	// silently filter the list against another resource's scope rows.
+	if filter.ResourceKind != domain.ResourceKindRelease {
+		t.Fatalf("filter kind = %q, want release", filter.ResourceKind)
+	}
```
Comment (IAM-marco) [inline internal/api/release_internal_test.go]: Simplification: this block re-tests shared machinery that already has its own test.

What changed: the new `TestReleaseListAccess` checks that an allowed call stamps the one-shot skip, that a narrowed call attaches the row filter, and that the filter carries the release kind.

Why this spot matters: the generic test already pins all of that machinery, for every resource at once:
https://github.com/zitadel/nextgen/blob/1b65bd691d683ea9178eb1a7b97cf972f7ebfead/internal/api/authz_internal_test.go#L370-L428
And the kind check on lines 105 to 109 cannot catch the one mistake it looks aimed at: this test passes `domain.ResourceKindRelease` into the call itself, so asserting the same value comes back out says nothing about what the real handler passes.

What follows concretely: about 27 lines of duplicated coverage that must be kept in sync with the shared test while adding no protection. No sibling resource repeats these machinery checks in its own test file.

Fix direction: drop lines 83 to 109 and keep only the release-flavored error checks below (lines 111 to 120), which are the one part the generic test cannot cover.

Coverage note: nothing is lost; the shared test already fails if the gate machinery breaks.

File: internal/api/release_internal_test.go [hunk 3]
```diff
@@ -60,3 +61,61 @@ func TestReleaseAccessRow(t *testing.T) {
...
+		Scope:         []string{"project.write", "project.read"},
+		PrincipalType: domain.AuthzPrincipalTypeSKProj,
+		PrincipalID:   "proj_a",
+	})
+	preview := WithScopeContext(context.Background(), ScopeContext{
+		ProjectID:     "proj_a",
+		Scope:         []string{"project.read"},
+		PrincipalType: domain.AuthzPrincipalTypeSKProj,
+		PrincipalID:   "proj_a",
+	})
```
Comment (IAM-marco) [inline internal/api/release_internal_test.go]: Simplification: this setup is copied word for word from `TestReleaseAccessRow` a few lines up in the same file:
https://github.com/zitadel/nextgen/blob/1b65bd691d683ea9178eb1a7b97cf972f7ebfead/internal/api/release_internal_test.go#L15-L29

Why this spot matters: if the machinery checks in the middle of this test go (see the other comment on this file), only the three release-flavored error checks on lines 111 to 120 survive, and they need exactly the setup that `TestReleaseAccessRow` already builds.

What follows concretely: folding those three checks into `TestReleaseAccessRow` as a subtest removes this whole function, its duplicated setup, and the `service` import that only the machinery checks use.

Fix direction: move lines 111 to 120 into `TestReleaseAccessRow` as a subtest and delete `TestReleaseListAccess`.

Coverage note: none needed; this only moves existing checks, it does not change what is covered.

File: internal/api/release_internal_test.go (hunk 2 above)
Comment (IAM-marco) [inline internal/api/release_internal_test.go]: Simplification: this block re-tests shared machinery that already has its own test.

What changed: the new `TestReleaseListAccess` checks that an allowed call stamps the one-shot skip, that a narrowed call attaches the row filter, and that the filter carries the release kind.

Why this spot matters: the generic test already pins all of that machinery, for every resource at once:
https://github.com/zitadel/nextgen/blob/d01e5655aed57ca0d6168c4e6fc67ece57361fea/internal/api/authz_internal_test.go#L370-L428
And the kind check on lines 105 to 109 cannot catch the one mistake it looks aimed at: this test passes `domain.ResourceKindRelease` into the call itself, so asserting the same value comes back out says nothing about what the real handler passes.

What follows concretely: about 27 lines of duplicated coverage that must be kept in sync with the shared test while adding no protection. No sibling resource repeats these machinery checks in its own test file.

Fix direction: drop lines 83 to 109 and keep only the release-flavored error checks below (lines 111 to 120), which are the one part the generic test cannot cover.

Coverage note: nothing is lost; the shared test already fails if the gate machinery breaks.

File: internal/api/release_internal_test.go (hunk 3 above)
Comment (IAM-marco) [inline internal/api/release_internal_test.go]: Simplification: this setup is copied word for word from `TestReleaseAccessRow` a few lines up in the same file:
https://github.com/zitadel/nextgen/blob/d01e5655aed57ca0d6168c4e6fc67ece57361fea/internal/api/release_internal_test.go#L15-L29

Why this spot matters: if the machinery checks in the middle of this test go (see the other comment on this file), only the three release-flavored error checks on lines 111 to 120 survive, and they need exactly the setup that `TestReleaseAccessRow` already builds.

What follows concretely: folding those three checks into `TestReleaseAccessRow` as a subtest removes this whole function, its duplicated setup, and the `service` import that only the machinery checks use.

Fix direction: move lines 111 to 120 into `TestReleaseAccessRow` as a subtest and delete `TestReleaseListAccess`.

Coverage note: none needed; this only moves existing checks, it does not change what is covered.

File: internal/api/release_internal_test.go (hunk 2 above)
Comment (IAM-marco) [reply internal/api/release_internal_test.go]: Please also use assert/require over `t.Fatal`

**Review by IAM-marco — CHANGES_REQUESTED** [review]

**Review by IAM-marco — APPROVED** [review]
Two-axis review ran on this branch: one correctness finding (the unmocked ReleaseService interface from the #1129 thread) and two test-duplication findings. All three were addressed on the branch (mock generated and wired into the unit tests, list gate folded into the access row test), so nothing is left to flag. Approving.
